### ISSUE TITLE
bf16 strided tensor wrong calculation

### ISSUE THREAD

**Issue by jay746** [pytorch/pytorch#83149]
### 🐛 Describe the bug

Issue when transpose is taken on bfloat16 tensor and passed to sum. the output is not correct. Example has been described below:
If you run same code in float32 then it works fine.
```
import torch
x = torch.ones([10, 13, 3, 3], dtype=torch.bfloat16)
x_trans = x.transpose(2, 3)
x_sum = torch.sum(x_trans, (0, 1, 2))
print(x_sum)
```

```
output: tensor([432., 432., 432.], dtype=torch.bfloat16)
but except output is tensor([390., 390., 390.])
```

### Versions

Collecting environment information...
PyTorch version: 1.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.10 (default, Jun 22 2022, 20:18:18)  [GCC 9.4.0] (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.3
[pip3] pytorch-lightning==1.6.5
[pip3] torch==1.12.0
[pip3] torch-tb-profiler==0.4.0
[pip3] torchmetrics==0.9.3
[pip3] torchvision==0.13.0
[conda] mkl                       2022.0.2                 pypi_0    pypi
[conda] mkl-include               2022.0.2                 pypi_0    pypi
[conda] numpy                     1.22.3                   pypi_0    pypi
[conda] pytorch-lightning         1.6.5                    pypi_0    pypi
[conda] torch                     1.12.0          pypi_0    pypi
[conda] torch-tb-profiler         0.4.0                    pypi_0    pypi
[conda] torchmetrics              0.9.3                    pypi_0    pypi
[conda] torchvision               0.13.0          pypi_0    pypi

cc @ezyang @gchanan @zou3519 @VitalyFedyunin @frank-wei

**Comment by jay746** [comment]
The same issue seen on permute and slice as well

**Comment by ngimel** [comment]
cc @peterbell10

**Comment by ngimel** [comment]
This is possibly a bf16 precision issue, if I reduce the size to (3, 13, 3, 3) I'm getting same results in both cases

**Comment by malfet** [comment]
@mingfeima can you please have a look at what is going on here?

**Comment by peterbell10** [comment]
I think the root cause here is a limitation of `TensorIterator` reductions. The non-transposed case is coalesced down to a 2-d operation whereas the transposed case becomes a 3d operation (two reduced dimensions and the middle non-reduced dim).

Since the cpu reduction loop only operates on two dimensions at a time, this means the transposed case is forced to do accumulation in the output and intermediate sums will be truncated to bfloat16 precision.

**Comment by mingfeima** [comment]
@CaoE , could you please help debug this issue?

**Comment by CaoE** [comment]
My observations are the same as @peterbell10's. `TensorIterator` coalesce the transposed case a 3d operation. 
For this case:
iter.shape : [3, 130, 3] iter.strides(0) : [0, 0, 2] iter.strides(1) :  [2, 18, 6].

The cpu reduction loop operates on two dimensions at a time, e.g. size0 is 3 and size1 is 130. 
Intermediate sums will be convert to bfloat16 precision:
```
struct CastStoreAccumulate {
  static void store(char * C10_RESTRICT data, int64_t stride, int64_t index, acc_t value) {
    auto * ptr = reinterpret_cast<scalar_t*>(data + index * stride);
    *ptr += value;
  }
};
```

**Comment by CaoE** [comment]
@peterbell10 `*ptr += value` will cause bf16 accumulation for such cases (TensorIterator coalesces tensors to >2 dims tensors, and the output stride is [0, 0, x, ...]).  It is hard to use acc buffer in the sum kernel since the outer data loop is controlled by  TensorIterator.

```
template <typename scalar_t, typename acc_t>
struct CastStoreAccumulate {
  static void store(char * C10_RESTRICT data, int64_t stride, int64_t index, acc_t value) {
    auto * ptr = reinterpret_cast<scalar_t*>(data + index * stride);
    *ptr += value;
  }
};
```
A simple idea is to directly replace the output with a float tensor during `TensorIterator` creation, and then copy the results back. But this will have a certain performance loss for all BF16 sums.

Another idea is to change `TensorIterator` and add an acc version: I have written a simple idea check version for this case https://github.com/pytorch/pytorch/pull/103956, which is not general enough. A general solution to this issue may require more changes to `TensorIterator`.

Could you please give some suggestions? Thank you.

**Comment by ezyang** [comment]
@peterbell10 As the one who implemented cascade reductions way back in the day, do you mind making the final call here?
